woocommerce/google-listings-and-ads
# Convert E2E tests from Puppeteer to Playwright

### Changes proposed in this Pull Request:

This PR converts the E2E tests from Puppeteer to Playwright and removes any dependencies on the `@woocommerce/e2e-*` packages. This is a followup from #2030

> **Note**
Until we resolve issue #2002 we are still using Node 14, which means the latest compatible version of Playwright was 1.33. Ideally this will be updated once we upgrade Node.

Overview of the changes to convert tests:
1. Conversion ID
The test Conversion ID used to be set through the Connection Test page. This has now been replaced with a custom plugin `tests/e2e/bin/test-data.php` which adds some API endpoints to set this data.

2. Site Settings
The payment gateway COD used to be enabled through an API call at the start of the test, this is now done in the `wp-env` start up script.

3. API requests
All API requests use the [basic auth plugin](https://github.com/WP-API/Basic-Auth) to send requests to the API. We use this for both WordPress and WooCommerce endpoints, see `tests/e2e/utils/api.js`

4. Avoiding `waitForSelector`
In our previous tests we often used `waitForSelector`, while this is still compatible with Playwright, I've replaced it with other [Locators](https://playwright.dev/docs/locators). This was done to simplify the tests and reduce the need for utility functions.

5. Mocking API requests
Instead of using a utility for mocking and spying on API requests, it's been replaced with the `route.fulfill`. We use this to mock connections to Jetpack or Google.

Closes #1599

### Detailed test instructions:

1. Checkout this PR and install packages: `nvm use && npm ci`
2. Start the `wp-env` environment: `npm run wp-env:up`
3. Install the Chromium browser for Playwright: `npx playwright install chromium`
4. Start the E2E tests: `npm run test:e2e`
5. Confirm that all the tests pass, and a test report is available in tests/e2e/test-result
6. Check that we can run the tests in debug mode and step through them in the browser: `npm run test:e2e-dev`

### Additional details:

Currently the E2E workflow is set to run only on release branches or manually. I've triggered a manual run, and all the tests passed: https://github.com/woocommerce/google-listings-and-ads/actions/runs/5753482975/job/15596778163

### Changelog entry
* Dev - Convert E2E tests from Puppeteer to Playwright.
